### PR TITLE
Restructured AICMessage for new tool_calls api; Updating messages via…

### PR DESCRIPTION
**Pull request by cielecki**
… websockets;

### CODE REVIEW

File: frontend/src/api/ws/handleChatMessage.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+                  ranCode = true;
+                }
+              }
+            }
+
+            if (!ranCode) {
+              await useChatStore.getState().doAnalysis();
+            }
+          }
+          break;
```
Comment (maciejt10c) [inline frontend/src/api/ws/handleChatMessage.ts:75]: can we extract this logic somehow to seperate file, switch in switch does not seems nice to read

Comment (maciejt10c) [reply frontend/src/api/ws/handleChatMessage.ts:75]: can we also do it for other cases, to enclose eg UpdateToolCallWsMessage in seperate function to minimalize size of the file to maximum

File: frontend/src/api/ws/handleChatMessage.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        case 'end':
+          if (data.analysis_request_id === useChatStore.getState().currentAnalysisRequestId) {
+            if (data.agent_id) useChatStore.setState({ agent_id: data.agent_id });
+            if (data.relevant_material_ids)
+              useChatStore.setState({ relevant_material_ids: data.relevant_material_ids });
+            if (data.next_step) useChatStore.setState({ next_step: data.next_step });
+            if (data.thinking_process) useChatStore.setState({ thinking_process: data.thinking_process });
+          }
+
```
Comment (maciejt10c) [inline frontend/src/api/ws/handleChatMessage.ts:189]: why do we need switch for that if every case is making same things, for `end` stage i would create just if? also it is more readable to use brackets for if statements {} :)

```typescript
      if (data.analysis_request_id === useChatStore.getState().currentAnalysisRequestId) {
            if (data.agent_id) useChatStore.setState({ agent_id: data.agent_id });
            if (data.relevant_material_ids)
              useChatStore.setState({ relevant_material_ids: data.relevant_material_ids });
            if (data.next_step) useChatStore.setState({ next_step: data.next_step });
            if (data.thinking_process) useChatStore.setState({ thinking_process: data.thinking_process });
      }       


       if (data.stage === 'end) {
          if (useChatStore.getState().agent_id !== 'user' && useChatStore.getState().next_step) {
            useChatStore.getState().appendGroup({
              agent_id: useChatStore.getState().agent_id || '',
              task: useChatStore.getState().next_step || '',
              materials_ids: useChatStore.getState().relevant_material_ids || [],
              role: 'assistant',
              messages: [],
            });

            useChatStore.getState().doExecute();
          }

          if (useChatStore.getState().currentAnalysisRequestId === data.analysis_request_id) {
            useChatStore.setState(() => ({
              currentAnalysisRequestId: undefined,
              agent_id: undefined,
              relevant_material_ids: undefined,
              next_step: undefined,
              thinking_process: undefined,
            }));
          }
}

```
btw: if statements with brackets are more readable, can we use it instead of one liners please

File: frontend/src/components/editables/chat/MessageGroup.tsx [hunk 3]
```diff
@@ -35,28 +28,10 @@ export function MessageGroup({
       )}
     >
       <div className="container flex mx-auto gap-5">
-        <UserInfo
-          agent_id={group.agent_id}
-          materials_ids={group.materials_ids}
-          task={group.task}
-        />
+        <UserInfo agent_id={group.agent_id} materials_ids={group.materials_ids} task={group.task} />
```
Comment (maciejt10c) [inline frontend/src/components/editables/chat/MessageGroup.tsx:31]: in js it is common to use a camelCase as a props naming convention

File: frontend/src/components/editables/chat/messages/MessageComponent.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+      saveCommandAndMessagesToHistory(content, group.role === 'user');
+    },
+    [message.id, editMessage, saveCommandAndMessagesToHistory, group.role],
+  );
+
+  const submitCommand = useChatStore((state) => state.submitCommand);
+
+  return (
+    <div key={message.id}>
```
Comment (maciejt10c) [inline frontend/src/components/editables/chat/messages/MessageComponent.tsx:43]: probably we dont need this key here anymore?

File: frontend/src/components/editables/chat/messages/ToolCall.tsx [hunk 5]
```diff
@@ -19,77 +19,71 @@ import { useCallback, useState } from 'react';
...
-  isStreaming: boolean;
+  tool_call: AICToolCall;
 }
 
-export function CodeMessage({ group, message, isStreaming }: MessageProps) {
-  const removeMessageFromGroup = useChatStore(
-    (state) => state.removeMessageFromGroup,
-  );
-  const editMessageContent = useChatStore((state) => state.editMessageContent);
+export function ToolCall({ group, tool_call }: MessageProps) {
```
Comment (maciejt10c) [inline frontend/src/components/editables/chat/messages/ToolCall.tsx:36]: prop name should in camelCase convention

File: frontend/src/store/editables/chat/ActionSlice.ts [hunk 6]
```diff
@@ -16,251 +16,84 @@
...
+      signal: get().executeAbortSignal.signal,
+    });
   },
 
   /**
    * doExecute expects that the last message is the one it should be filling in.
    */
   doExecute: async () => {
-
+    console.log('Executing');
```
Comment (maciejt10c) [inline frontend/src/store/editables/chat/ActionSlice.ts:74]: leftover?

File: frontend/src/store/editables/chat/CommandSlice.ts [hunk 7]
```diff
@@ -18,22 +18,19 @@ import { StateCreator } from 'zustand';
...
 
 export type CommandSlice = {
   commandHistory: string[];
   commandIndex: number;
   historyUp: () => void;
   historyDown: () => void;
-  scrollChatToBottom: undefined | ((props: {behavior: "auto" | "smooth" | undefined}) => void);
-  setScrollChatToBottom: (scrollChatToBottom: (props: {behavior: "auto" | "smooth" | undefined}) => void) => void;
+  scrollChatToBottom: undefined | ((props: { behavior: 'auto' | 'smooth' | undefined }) => void);
+  setScrollChatToBottom: (scrollChatToBottom: (props: { behavior: 'auto' | 'smooth' | undefined }) => void) => void;
```
Comment (maciejt10c) [inline frontend/src/store/editables/chat/CommandSlice.ts:29]: can you assign this prop type to Type to not repeat it twice

File: frontend/src/store/editables/chat/MessageSlice.ts [hunk 8]
```diff
@@ -137,85 +114,82 @@ export const createMessageSlice: StateCreator<ChatStore, [], [], MessageSlice> =
...
-
-    get().saveCurrentChatHistory();
   },
-  appendEmptyOutput: (groupId?: string, messageId?: string) => {
+  appendToolCall: (toolCall: AICToolCall, messageId?: string) => {
     set((state) => {
-      if (!state.chat) { throw new Error('Chat is not initialized'); }
+      if (!state.chat) {
+        throw new Error('Chat is not initialized');
+      }
```
Comment (maciejt10c) [inline frontend/src/store/editables/chat/MessageSlice.ts:170]: can you make a helper for that, because i saw we are using it in a few places

Comment (maciejt10c) [reply frontend/src/store/editables/chat/MessageSlice.ts:170]:
```typescript 
const checkIfHasChat = (chat: Chat) => {
     if (!state.chat) {
        throw new Error('Chat is not initialized');
      }
}
```

File: frontend/src/utils/editables/chatUtils.ts [hunk 9]
```diff
@@ -48,64 +50,105 @@ export function getGroup(
...
+          messageIndex,
+          message,
+        };
+      }
+
+      messageIndex++;
+    }
+    groupIndex++;
   }
+}
```
Comment (maciejt10c) [inline frontend/src/utils/editables/chatUtils.ts:87]: this looks a bit hard to read can we simplify this a bit

```typescript
export function getMessage(chat, messageId) {
  const groupIndex = chat.message_groups.findIndex(group =>
    group.messages.some(message => message.id === messageId)
  );

  if (groupIndex !== -1) {
    const group = chat.message_groups[groupIndex];
    const messageIndex = group.messages.findIndex(message => message.id === messageId);
    const message = group.messages[messageIndex];
    return { groupIndex, group, messageIndex, message };
  }
}
```

File: frontend/src/utils/editables/chatUtils.ts [hunk 10]
```diff
@@ -48,64 +50,105 @@ export function getGroup(
...
+        }
+        outputIndex++;
+      }
+      messageIndex++;
+    }
+    groupIndex++;
+  }
+
+  return undefined;
+}
```
Comment (maciejt10c) [inline frontend/src/utils/editables/chatUtils.ts:138]:
```typescript
export function getToolCall(chat, outputId) {
  const foundToolCall = chat.message_groups.reduce((result, group, groupIndex) => {
    if (result) return result;

    const messageWithToolCall = group.messages.find((message, messageIndex) => {
      const toolCallIndex = message.tool_calls.findIndex(toolCall => toolCall.id === outputId);
      if (toolCallIndex !== -1) {
        return {
          groupIndex,
          group,
          messageIndex,
          message,
          toolCallIndex,
          toolCall: message.tool_calls[toolCallIndex],
        };
      }
    });

    return messageWithToolCall;
  }, null);

  return foundToolCall || undefined;
}
```

Comment (cielecki) [reply frontend/src/utils/editables/chatUtils.ts:138]: Is this actually better?

Comment (maciejt10c) [reply frontend/src/utils/editables/chatUtils.ts:138]: i am not sure if this is better but imo more readable @aleksanderfret what do you think about this ?

File: frontend/src/api/ws/handleChatMessage.ts (hunk 1 above)
Comment (aleksanderfret) [reply frontend/src/api/ws/handleChatMessage.ts:75]: I would also try to create a separate helper for nested switch statements

File: frontend/src/api/ws/useWebSocketStore.ts [hunk 11]
```diff
@@ -145,6 +137,15 @@ export const useWebSocketStore = create<WebSockeStore>((set, get) => ({
             });
           }
           break;
+        case 'ResetMessageWSMessage':
+        case 'UpdateAnalysisWSMessage':
+        case 'UpdateMessageWSMessage':
+        case 'UpdateToolCallWSMessage':
+        case 'UpdateToolCallOutputWSMessage':
+          await handleChatMessage(data);
```
Comment (aleksanderfret) [inline frontend/src/api/ws/useWebSocketStore.ts:145]: We have one huge callback that handles all above Messages. Can we create a message dedicated handlers to make them smaller and easier to follow?

File: frontend/src/components/editables/chat/messages/MessageComponent.tsx (hunk 4 above)
Comment (aleksanderfret) [reply frontend/src/components/editables/chat/messages/MessageComponent.tsx:43]: Is this key for forcing re-render on message changes?

File: frontend/src/components/editables/chat/messages/MessageComponent.tsx [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        handleRemoveClick={handleRemoveClick}
+      >
+        <div className="flex flex-col gap-2">
+          {(message.content || message.is_streaming) && (
+            <>
+              {group.role !== 'user' && (
+                <div className="flex-grow">
+                  <div className="prose prose-stone dark:prose-invert ">
+                    <ReactMarkdown
+                      components={{
```
Comment (aleksanderfret) [inline frontend/src/components/editables/chat/messages/MessageComponent.tsx:57]: Can we extract Markdown to a separate component and pass required props?

File: frontend/src/utils/editables/chatUtils.ts (hunk 10 above)
Comment (aleksanderfret) [reply frontend/src/utils/editables/chatUtils.ts:138]: TBH, having no idea what this code is about it easier for me to understand  @cielecki approach even in the nested loops. @maciejt10c Without any feature-related knowledge using reduce and related logic it hard to follow what is going on there.

File: frontend/src/store/editables/chat/MessageSlice.ts (hunk 8 above)
Comment (cielecki) [reply frontend/src/store/editables/chat/MessageSlice.ts:170]: Typechecking will not work with this solution

Comment (cielecki) [reply frontend/src/store/editables/chat/MessageSlice.ts]: Fixed here: https://github.com/10clouds/aiconsole/commit/f5698b987cc11810dc11de6c56ba8fa2ce58c60e

File: frontend/src/components/editables/chat/MessageGroup.tsx (hunk 3 above)
Comment (cielecki) [reply frontend/src/components/editables/chat/MessageGroup.tsx]: https://github.com/10clouds/aiconsole/commit/26c4285bd2f1260c811770909207858c788c3b06

File: frontend/src/api/ws/handleChatMessage.ts [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+          break;
+        }
+      }
+      break;
+    case 'UpdateAnalysisWSMessage':
+      switch (data.stage) {
+        case 'start':
+          if (data.analysis_request_id === useChatStore.getState().currentAnalysisRequestId) {
+            if (data.agent_id) useChatStore.setState({ agent_id: data.agent_id });
```
[the remark this answers is not in the document]
Comment (cielecki) [reply frontend/src/api/ws/handleChatMessage.ts]: Fixed

File: frontend/src/api/ws/useWebSocketStore.ts (hunk 11 above)
Comment (cielecki) [reply frontend/src/api/ws/useWebSocketStore.ts:145]: https://10clouds.atlassian.net/browse/ACD-663

File: frontend/src/utils/editables/chatUtils.ts (hunk 9 above)
Comment (cielecki) [reply frontend/src/utils/editables/chatUtils.ts]: https://github.com/10clouds/aiconsole/commit/59c0a9ba593d7906d4dbc57ca40d6c4b9631a3b2

File: frontend/src/components/editables/chat/messages/MessageComponent.tsx (hunk 4 above)
Comment (cielecki) [reply frontend/src/components/editables/chat/messages/MessageComponent.tsx:43]: Fixed, old code

Comment (cielecki) [reply frontend/src/components/editables/chat/messages/MessageComponent.tsx]: https://github.com/10clouds/aiconsole/commit/ca4781b7c36ae21e264b5b98b92fa391d4afe30e

File: frontend/src/utils/editables/chatUtils.ts (hunk 10 above)
Comment (cielecki) [reply frontend/src/utils/editables/chatUtils.ts]: I'm leaving this as it is

File: frontend/src/utils/editables/chatUtils.ts (hunk 9 above)
Comment (cielecki) [reply frontend/src/utils/editables/chatUtils.ts]: Leaving this as it is.

File: frontend/src/components/editables/chat/messages/ToolCall.tsx (hunk 5 above)
Comment (cielecki) [reply frontend/src/components/editables/chat/messages/ToolCall.tsx]: Fixed in https://github.com/10clouds/aiconsole/commit/14122140bb8d5c6b1243ad26f51ef332874722d3
